### PR TITLE
Improvement/error code

### PR DESCRIPTION
**Pull request by rajeev-0**

### CODE REVIEW

File: cmp_doc/HOWTO.md [hunk 1]
```diff
@@ -202,4 +202,5 @@ for the input field with label "Serial port:", select, e.g., "COM5").
 		- HTTP or CoAP as transport
 - For UART terminal based 'CLI':
 	- Follow the instructions on the terminal
+	- Make sure inbound firewall rules is allowed for LRA running on PC, Otherwise client will not be able to communicate with LRA.
```
Comment (DDvO) [inline cmp_doc/HOWTO.md]: -> Make sure that on a Windows host inbound messages to the CMP server (e.g., LRA) application are allowed by the firewall rules. Otherwise the client will not be able to communicate with the server.

File: cmplib/cmpcl.h [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
 #ifndef CMPCL_H
 #define CMPCL_H
-
+#include "cmpcl_error.h"
```
Comment (DDvO) [inline cmplib/cmpcl.h:27]: Please retain the empty line before the `#include`.

File: cmplib/cmpcl_ctx.c [hunk 3]
```diff
@@ -73,11 +73,11 @@ int cmp_ctx_init(cmp_ctx *ctx, mbedtls_ctr_drbg_context *ctr_drbg,
 	ctx->send_receive_cb = send_receive_func;
 
 	if ((ctx->shost = strdup(shost)) == NULL) {
-		return -1; /* TODO: improve error code */
+		return CMPCL_ERR_HOST_IP;
```
Comment (DDvO) [inline cmplib/cmpcl_ctx.c]: Here and in the two similar cases below, `CMPCL_ERR_MEMORY_ALLOCATION` would be sufficient.
(Yet if you choose to be more specific, better use, e.g., `CMPCL_ERR_HOST_STR_MEMORY_ALLOCATION`.)

File: cmplib/cmpcl_error.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#define CMPCL_ERR_ASN1_PARSING			-0x000A		/** Error in parsing ASN1  */
+#define CMPCL_ERR_CERT_OR_ENCCERT		-0x0010		/** Invalid CertOrEncCert choice */
+#define CMPCL_ERR_POPO_METHOD			-0x0011 	/** Invalid/Unsupported POPO Method */
+#define CMPCL_ERR_KUR_OLDCERT			-0x0012 	/** Existing certificate missing for KUR  */
+#define CMPCL_ERR_CERTREP_B_PARSING		-0x0013  	/** Error in parsing certrep body */
+#define CMPCL_ERR_WRONG_RESP_TYPE		-0x0014		/** Wrong response type received  */
+#define CMPCL_ERR_UNSUPPORTED_BODYTYPE	-0x0015	    /** Received body type not supported  */
+#define CMPCL_ERR_CERT_NOT_RECEIVED		-0x0016	 	/** No certificate received in certrep message */
+#define CMPCL_ERR_FILE_OPEN				-0x0017	 	/** Error in file opening */
+#define CMPCL_ERR_FILE_WRITE			-0x0018		/** Error in file writing */
```
Comment (DDvO) [inline cmplib/cmpcl_error.h]: Please make sure using space chars that the error code and comment columns are aligned.

File: cmplib/cmpcl_error.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/*
+ * cmpcl_error.h
+ *
+ *  Created on: Aug 16, 2022
+ *      Author: z004hm2h
+ */
```
Comment (DDvO) [inline cmplib/cmpcl_error.h]: Please use a file header including copyright and license statements like in the other source files here, e.g., `cmp_main.h`.

File: cmplib/cmpcl_lib.c [hunk 6]
```diff
@@ -90,7 +90,7 @@ int setStr( unsigned char **dst, const unsigned char *src, const size_t len)
         mbedtls_free( *dst);
     if( (*dst = (unsigned char*) mbedtls_calloc(1, len)) == NULL) {
         CMPERRS("Error allocating space\n");
-        exit(1); /* TODO: handle better */
+        exit(CMPCL_ER_MEMORY_ALLOCATION); /* TODO: handle better */
```
Comment (DDvO) [inline cmplib/cmpcl_lib.c]: `_ER_` -> `_ERR_` - otherwise I wonder how this compiles successfully?
And I'm not sure if it is adequate to use an error code as an exit code.
Likely better is to throw an error and return the error code as regular result, which of course needs to be handled by the caller.

File: cmplib/cmpcl_ctx.c [hunk 7]
```diff
@@ -136,14 +136,13 @@ int cmp_ctx_set_reference(cmp_ctx *ctx, const unsigned char *ref,
 /* **************************************************************** */
 int cmp_ctx_set_messageTime(cmp_ctx *ctx, const char *msgTime) {
 	if (strlen(msgTime) != MBEDTLS_X509_RFC5280_UTC_TIME_LEN - 1) {
-		return ( MBEDTLS_ERR_X509_BAD_INPUT_DATA); /* TODO this should be changed?  */
+		return ( CMPCL_ERR_MSGTIME_LEN);
```
Comment (DDvO) [inline cmplib/cmpcl_ctx.c]: Better not use needless parentheses, here and in various further places (even though they were used in the pre-existing code for whatever reason).

File: cmplib/cmpcl_read.c [hunk 8]
```diff
@@ -55,6 +57,12 @@ static void zeroize( void *v, size_t n ) {
 
 #define PARSE_ASN1_CONSTRUCTED_TAG(expected) ((ret = mbedtls_asn1_get_tag( &p, end, &len, MBEDTLS_ASN1_CONSTRUCTED | expected ) ) != 0)
 
+#define MBEDTLS_ERR(ret, err_buf, CMPCL_ER_BUF_LEN)  \
+			do{\
```
Comment (DDvO) [inline cmplib/cmpcl_read.c:61]: Better move the below declaration `char err_buf[CMPCL_ER_BUF_LEN];` in here and remove the `err_buf` parameter, such that the caller does not need to bother with it.

File: cmplib/cmpcl_read.c [hunk 9]
```diff
@@ -703,11 +714,19 @@ static int cmp_pkibody_PKIStatusInfo_parse_der( cmp_PKIStatusInfo *sinfo,
...
-                sinfo->statusString->buf.p);
+        {
+        	MBEDTLS_ERR( ret, err_buf, CMPCL_ER_BUF_LEN );
+        	return( CMPCL_ERR_ASN1_PARSING );
+
+        }
+        mbedtls_asn1_sequence *StatusString = sinfo->statusString;
+        while(StatusString)
+        {
+        	CMPERRV("Response statusString: %.*s",
```
Comment (DDvO) [inline cmplib/cmpcl_read.c]: Please make sure that the "Response statusString: " part is printed only once, no matter how many iterations are in the loop.

File: cmplib/cmpcl_read.c [hunk 10]
```diff
@@ -703,11 +714,19 @@ static int cmp_pkibody_PKIStatusInfo_parse_der( cmp_PKIStatusInfo *sinfo,
...
+                (int) StatusString->buf.len,
+				StatusString->buf.p);
+        	StatusString = StatusString->next;
+        }
```
Comment (DDvO) [inline cmplib/cmpcl_read.c]: I'd prefer using a `for(...)` loop.

File: cmplib/cmpcl_read.c [hunk 11]
```diff
@@ -676,8 +685,10 @@ static int cmp_pkibody_PKIStatusInfo_parse_der( cmp_PKIStatusInfo *sinfo,
...
     if( ( ret = mbedtls_asn1_get_int( &p, end, &sinfo->PKIStatus ) ) != 0 )
-        return( ret ); /* TODO: improve */
-
+    {
+    	MBEDTLS_ERR( ret, err_buf, CMPCL_ER_BUF_LEN );
+        return( CMPCL_ERR_ASN1_PARSING );
+    }
     /* not accepted ? */
     int ok = sinfo->PKIStatus == CMP_PKISTATUS_ACCEPTED;
     if(ok) {
```
Comment (DDvO) [inline cmplib/cmpcl_read.c:719]: Please change the following (existing) line to:
```
CMPDBGV("Response with PKIStatus: CMP_PKISTATUS_ACCEPTED");
```

File: cmplib/cmpcl_read.c [hunk 12]
```diff
@@ -971,15 +993,61 @@ static int cmp_pkibody_errmsgcnt_parse_der( cmp_ErrorMsgContent *emc,
...
+	   }
+	   mbedtls_asn1_sequence *ErrorDetails = emc->errorDetails;
+	   while(ErrorDetails)
+	   {
+		CMPERRV("Response Error Details: %.*s",
+			   (int) ErrorDetails->buf.len,
+			   ErrorDetails->buf.p);
+		ErrorDetails = ErrorDetails->next;
+	   }
+	}
```
Comment (DDvO) [inline cmplib/cmpcl_read.c]: To avoid redundancy, better factor out this code section and essentially the same code in `cmp_pkibody_PKIStatusInfo_parse_der` into an auxiliary function that handles parsing and printing of `PKIFreeText` in a generic way.

File: cmplib/cmpcl_read.c [hunk 13]
```diff
@@ -994,6 +1062,13 @@ int cmp_pkimessage_parse_check_der(cmp_ctx *ctx, int expected_type, cmp_pkimessa
     unsigned char *p, *end;
 
     CMPINFOV("----"); // end of transmission
+#ifdef DEBUG_ASN1
+    CMPINFOV("\r\n **Reply** \r\n");
+    int i;
+    for(i = 0; i < buflen+1; i++)
```
Comment (DDvO) [inline cmplib/cmpcl_read.c]: Are you sure the `+1` is correct here? This prints one more byte than `buflen`.

File: cmplib/cmpcl_read.c [hunk 14]
```diff
@@ -1253,8 +1327,8 @@ int cmp_pkimessage_parse_check_der(cmp_ctx *ctx, int expected_type, cmp_pkimessa
         }
 
         if( ( ret = cmp_x509_crt_verify(ctx->extraCerts, //mbedtls_x509_crt *crt,
-                ctx->prot_trust_anchor, //mbedtls_x509_crt *trust_ca,
-                ctx->prot_crls, //mbedtls_x509_crl *ca_crl,
+                ctx->enrol_trust_anchor, //mbedtls_x509_crt *trust_ca,
+                ctx->enrol_crls, //mbedtls_x509_crl *ca_crl,
```
Comment (DDvO) [inline cmplib/cmpcl_read.c]: Why did you change this?
This looks wrong because at this point the (general) message protection is checked for any kind of message,
not the cert returned in a positive ip/cp/kup. Yet for the latter situation, would be nice to verify the received new cert in addition, using the `enrol_{trust_anchor,enrol_crtls}`, in `cmpcl_do_transaction()`.

File: cmplib/cmpcl_ses.c [hunk 15]
```diff
@@ -65,7 +65,7 @@ int cmpcl_do_transaction(cmp_ctx *ctx, const int body_type) {
 
 	if (len < 0) {
 		CMPERRV("cmpcl_CMPwrite_PKIMessage der returned %d", len);
-		ret =  -1; /* TODO: improve error code? */
+		ret = len; /* updating error code*/
```
Comment (DDvO) [inline cmplib/cmpcl_ses.c]: `updating error code*/` ->
`if negative, len is an error code */`

File: cmplib/cmpcl_ses.c [hunk 16]
```diff
@@ -156,7 +157,7 @@ int cmpcl_do_transaction(cmp_ctx *ctx, const int body_type) {
 
 	    if (len < 0) {
 	        CMPERRV("cmpcl_CMPwrite_PKIMessage der returned %d", len);
-	        ret = -1; /* TODO: improve error code? */
+	        ret = len; /* updating error code */
```
Comment (DDvO) [inline cmplib/cmpcl_ses.c]: `updating error code` ->
`if negative, len is an error code`

File: cmplib/cmpcl_write.c [hunk 17]
```diff
@@ -853,6 +853,14 @@ int cmpcl_CMPwrite_PKIMessage_der( cmp_ctx *ctx, unsigned char *start,
     write_to_file("PKIMessage.der", *p, len);
 #endif
 
+#ifdef DEBUG_ASN1
+    int i;
+    CMPINFOV("\r\n **Request** \r\n");
+    for(i = 0; i < len+1; i++)
```
Comment (DDvO) [inline cmplib/cmpcl_write.c]: Is the  `+1` really correct?

**Review by DDvO — CHANGES_REQUESTED** [review]
Nice improvements.
Still I have a couple of (mostly very minor) change requests.

File: cmplib/cmpcl_lib.c (hunk 6 above)
Comment (rajeev-0) [reply cmplib/cmpcl_lib.c]: It is compiling because "exit" itself is macro with no reference to its argument.
Removed exit to throw error.

File: cmplib/cmpcl_ctx.c (hunk 3 above)
Comment (DDvO) [reply cmplib/cmpcl_ctx.c]: Well, if you use such a specific code `CMPCL_ERR_HOST_STR_MEMORY_ALLOCATION`, the other codes should be `CMPCL_ERR_PATH_STR_MEMORY_ALLOCATION` etc. 
Or simpler, as mentioned before, `CMPCL_ERR_MEMORY_ALLOCATION` in all three places.

File: cmplib/cmpcl_lib.c (hunk 6 above)
Comment (DDvO) [reply cmplib/cmpcl_lib.c]: I see.

File: cmplib/cmpcl_read.c (hunk 8 above)
Comment (DDvO) [reply cmplib/cmpcl_read.c:61]: Hmm, moving the decl to file-level was not what I meant - this might cause trouble with multi-threaded invocation.
Please move the decl inside the `while` loop.

**Review by DDvO — APPROVED** [review]
Thanks - looks good
